graph/articulationpoints.go

```diff
@@ -0,0 +1,58 @@
+// Algorithm to identify articulation points in a connected graph
+// You can read more about articulation points here: https://en.wikipedia.org/wiki/Biconnected_component
+// and the implementation and explanation here: https://cptalks.quora.com/Cut-Vertex-Articulation-point
+
+package graph
+
+import "github.com/TheAlgorithms/Go/math/min"
+
+var (
+	time               int    // time variable to keep track of the time of discovery_time of a vertex
+	is_ap              []bool // boolean array to mark articulation points
+	visited            []bool // boolean array to mark visited vertices
+	child_cnt          []int  // array to store the number of children of a vertex
+	discovery_time     []int  // array to store the time of discovery of a vertex
+	earliest_discovery []int  // array to store the earliest discovered vertex reachable from a vertex
+)
+
+func ArticulationPoint(graph Graph) []bool {
+	//initialize all the variables
+	time = 0
+	is_ap = make([]bool, graph.vertices)
+	visited = make([]bool, graph.vertices)
+	child_cnt = make([]int, graph.vertices)
+	discovery_time = make([]int, graph.vertices)
+	earliest_discovery = make([]int, graph.vertices)
+
+	dfs(graph, 0, -1)
+
+	if child_cnt[0] == 1 {
+		is_ap[0] = false // if the root has only one child, it is not an articulation point
+	}
+
+	return is_ap
+}
+
+func dfs(graph Graph, u int, parent int) {
```

Comment:
There is already a depth first search algorithm within this repository - in fact within this package. I recommend that you stick to using the ones that are already there.

Reply:
@tjgurwara99 so the depth first search implemented here is a bit different from the dfs defined in the graphs folder. For instance, as part of the dfs I used for this use-case, I am assigning and updating the `discovery_time` of the vertices which is not part of the dfs already implemented, moreover manipulating other variables like `earliest_discovery`, `time`, `child` and `parent` are again not included in a typical dfs. Either I go and tweak the dfs implementation and include all these extraneous parameters or should I chnage the number of the function to something like `ArticulationPointHelper` to avoid confusion with a normal dfs?

Reply:
> @tjgurwara99 so the depth first search implemented here is a bit different from the dfs defined in the graphs folder.

Exactly why I mentioned it here, calling it `dfs` is not ideal because dfs doesn't change any state in the graph.

> Either I go and tweak the dfs implementation and include all these extraneous parameters or should I change the name of the function to something like `ArticulationPointHelper` to avoid confusion with a normal dfs?

No you shouldn't tweak the already implemented dfs as your implementation is not purely dfs but rather an application of it. I suggest to call this `articulationPointHelper` and including a documentation comment of why its needed and its basis (based on dfs in this case).

Reply:
@tjgurwara99 sure, I'll do that.